### ISSUE TITLE
Perl 5.24 makes nama FTBFS due to segfault

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#15930]
Migrated from [rt.perl.org#131054](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=131054) (status was 'open')

Searchable as RT131054$

**Comment by p5pRT** [comment]
### From balint.reczey@canonical.com


This is a bug report for perl from rbalint@&#8203;ubuntu\.com\,
generated with the help of perlbug 1\.40 running under perl 5\.24\.1\.




---
When building nama Perl crashes in tests\.

Originally reported to Debian BTS&#8203;:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=839218

On Fri\, 30 Sep 2016 10&#8203;:09&#8203;:01 \+0100 Chris Lamb \<lamby@&#8203;debian\.org> wrote&#8203;:
> Source&#8203;: nama
> Version&#8203;: 1\.208\-1
> Severity&#8203;: serious
> Justification&#8203;: fails to build from source
> User&#8203;: reproducible\-builds@&#8203;lists\.alioth\.debian\.org
> Usertags&#8203;: ftbfs
> X\-Debbugs\-Cc&#8203;: reproducible\-builds@&#8203;lists\.alioth\.debian\.org
> 
> Dear Maintainer\,
> 
> nama fails to build from source in unstable/amd64&#8203;:
> 
>   \[\.\.\]
> 
>   ok 57 \- set JACK client as input
>   sax&#8203;: JACK source port is sax\_in\. Make connections manually\.
>   
>   Track sax&#8203;: source set to JACK manual port sax\_in
>   Failed 69/126 subtests 
>   t/13\_io\.t \.\.\.\.\.\.\. 
>   1\.\.1
>   ok 1 \- use Audio&#8203;::Nama&#8203;::IO;
>   ok
>   
>   Test Summary Report
>   \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>   t/12\_nama\.t   \(Wstat&#8203;: 139 Tests&#8203;: 57 Failed&#8203;: 0\)
>     Non\-zero wait status&#8203;: 139
>     Parse errors&#8203;: Bad plan\.  You planned 126 tests but ran 57\.
>   Files=7\, Tests=91\,    5 wallclock secs \( 0\.03 usr  0\.01 sys \+  2\.29 cusr  0\.43 csys =  2\.76 CPU\)
>   Result&#8203;: FAIL
>   Failed 1/7 test programs\. 0/91 subtests failed\.
>   Makefile&#8203;:930&#8203;: recipe for target 'test\_dynamic' failed
>   make\[1\]&#8203;: \*\*\* \[test\_dynamic\] Error 255
>   make\[1\]&#8203;: Leaving directory '/home/lamby/temp/cdt\.20160930095107\.55GhkE5sL3\.db\.nama/nama\-1\.208'
>   dh\_auto\_test&#8203;: make \-j1 test TEST\_VERBOSE=1 returned exit code 2
>   debian/rules&#8203;:4&#8203;: recipe for target 'build' failed
>   make&#8203;: \*\*\* \[build\] Error 2
> 
>   \[\.\.\]


This looks like a problem in Perl itself possibly causing random crashes
elsewhere\, too\.

Reproduction with the perl\-debug package shows the internal assertion&#8203;:

sudo apt\-get install perl\-debug
sudo apt\-get build\-dep nama
apt source nama
cd nama\-\*
dpkg\-buildpackage
\.\.\.
ok 57 \- set JACK client as input
sax&#8203;: JACK source port is sax\_in\. Make connections manually\.

Track sax&#8203;: source set to JACK manual port sax\_in
Failed 69/126 subtests
t/13\_io\.t \.\.\.\.\.\.\.
1\.\.1
ok 1 \- use Audio&#8203;::Nama&#8203;::IO;
ok
\.\.

PERL\_DL\_NONLAZY=1 PERL\_USE\_UNSAFE\_INC=1 "/usr/bin/debugperl" \\
"\-MExtUtils&#8203;::Command&#8203;::MM" "\-MTest&#8203;::Harness" "\-e" " test\_harness\(1\, \\
'inc'\, 'blib/lib'\, 'blib/arch'\)" t/12\_nama\.t

\.\.\.

ok 57 \- set JACK client as input
sax&#8203;: JACK source port is sax\_in\. Make connections manually\.
Track sax&#8203;: source set to JACK manual port sax\_in
debugperl&#8203;: sv\.c&#8203;:6438&#8203;: Perl\_sv\_clear&#8203;: Assertion \`SvTYPE\(sv\) \!=
\(svtype\)SVTYPEMASK' failed\.
Failed 69/126 subtests
\.\.\.

Address Sanitizer shows the crash with non\-debug perl&#8203;:

LD\_PRELOAD=libasan\.so\.3 PERL\_DL\_NONLAZY=1 PERL\_USE\_UNSAFE\_INC=1 \\
"/usr/bin/perl" "\-MExtUtils&#8203;::Command&#8203;::MM" "\-MTest&#8203;::Harness" "\-e" "
test\_harness\(1\, 'inc'\, 'blib/lib'\, 'blib/arch'\)" t/12\_nama\.t
\.\.\.
sax&#8203;: JACK source port is sax\_in\. Make connections manually\.
Track sax&#8203;: source set to JACK manual port sax\_in
ASAN&#8203;:DEADLYSIGNAL
=================================================================
==13877==ERROR&#8203;: AddressSanitizer&#8203;: SEGV on unknown address 0x00ff00000012
\(pc 0x559191b15fd8 bp 0x62100013f578 sp 0x7fffaa838b00 T0\)
&nbsp;    \#0 0x559191b15fd7  \(/usr/bin/perl\+0xd9fd7\)
&nbsp;    \#1 0x559191b1616d in Perl\_sv\_unmagic \(/usr/bin/perl\+0xda16d\)
&nbsp;    \#2 0x559191b1513a in Perl\_sv\_clear \(/usr/bin/perl\+0xd913a\)
&nbsp;    \#3 0x559191b15a8f in Perl\_sv\_free2 \(/usr/bin/perl\+0xd9a8f\)
&nbsp;    \#4 0x559191b4624e in Perl\_leave\_scope \(/usr/bin/perl\+0x10a24e\)
&nbsp;    \#5 0x559191b49707  \(/usr/bin/perl\+0x10d707\)
&nbsp;    \#6 0x559191b4ebab in Perl\_die\_unwind \(/usr/bin/perl\+0x112bab\)
&nbsp;    \#7 0x559191aea518 in Perl\_vcroak \(/usr/bin/perl\+0xae518\)
&nbsp;    \#8 0x559191aeafb3 in Perl\_croak \(/usr/bin/perl\+0xaefb3\)
&nbsp;    \#9 0x559191b1d592 in Perl\_sv\_setsv\_flags \(/usr/bin/perl\+0xe1592\)
&nbsp;    \#10 0x559191b08ca9 in Perl\_pp\_sassign \(/usr/bin/perl\+0xccca9\)
&nbsp;    \#11 0x559191b08605 in Perl\_runops\_standard \(/usr/bin/perl\+0xcc605\)
&nbsp;    \#12 0x559191a8e6c8 in perl\_run \(/usr/bin/perl\+0x526c8\)
&nbsp;    \#13 0x559191a6787c in main \(/usr/bin/perl\+0x2b87c\)
&nbsp;    \#14 0x7fa31be502b0 in \_\_libc\_start\_main
\(/lib/x86\_64\-linux\-gnu/libc\.so\.6\+0x202b0\)
&nbsp;    \#15 0x559191a678b9 in \_start \(/usr/bin/perl\+0x2b8b9\)

AddressSanitizer can not provide additional info\.
SUMMARY&#8203;: AddressSanitizer&#8203;: SEGV \(/usr/bin/perl\+0xd9fd7\)
==13877==ABORTING
Dubious\, test returned 1 \(wstat 256\, 0x100\)
Failed 69/126 subtests


Niko Tyni found a workaround\, but Perl itself would still be buggy&#8203;:

> While I haven't made it to the bottom of this yet\, I note that there's
> this code in lib/Audio/Nama/Effect\.pm&#8203;:
>
>          my $perl\_version = $^V;
>          my \($minor\_version\) = $perl\_version =~ /^v5\\\.\(\\d\+\)/;
>          my @&#8203;new\_list = grep  \{ $\_ ne $id  \} @&#8203;ops\_list;
>          \#say "new\_list&#8203;: @&#8203;new\_list";
>          if \($minor\_version \<= 14\)
>               \{    $track\->\{ops\}   = \[ @&#8203;new\_list  \] \}
>          else \{ @&#8203;\{ $track\->\{ops\} \} =    @&#8203;new\_list    \}
>
> and if I patch that to take the \<= 14 code path\, the segfault goes away
> here and all tests pass\. What's the history behind this version check?

Funny story\, it was added to work around a segfault in older perl&#8203;:
https://github.com/bolangi/nama/commit/8ecdbd5ad0b8ac67b7e4adab2ee1100fbf29aa1f

The line looks very much like a corner case for the assignment
reimplementation&#8203;:


 https://github.com/Perl/perl5/commit/a5f48505593c7e1ca478de383e24d5cc2541f3ca

&nbsp;       re\-implement OPpASSIGN\_COMMON mechanism

 This commit almost completely replaces the current mechanism
 for detecting and handing common vars in list assignment\, e\.g\.
 \.\.\.


The issue was also reported against nama\, but I could not find a way
to reassign it to perl&#8203;:
https://rt.cpan.org/Public/Bug/Display.html?id=109852




---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=high


---
Site configuration information for perl 5\.24\.1&#8203;:

Configured by Debian Project at Thu Mar 16 11&#8203;:11&#8203;:46 UTC 2017\.

Summary of my perl5 \(revision 5 version 24 subversion 1\) configuration&#8203;:
&nbsp;     Platform&#8203;:
&nbsp;    osname=linux\, osvers=3\.16\.0\, archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux localhost 3\.16\.0 \#1 smp debian 3\.16\.0 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dcc=x86\_64\-linux\-gnu\-gcc \-Dcpp=x86\_64\-linux\-gnu\-cpp \-Dld=x86\_64\-linux\-gnu\-gcc \-Dccflags=\-DDEBIAN \-Wdate\-time \-D\_FORTIFY\_SOURCE=2 \-g \-O2 \-fdebug\-prefix\-map=/build/perl\-fHa8XZ/perl\-5\.24\.1=\. \-fstack\-protector\-strong \-Wformat \-Werror=format\-security \-Dldflags= \-Wl\,\-z\,relro \-Dlddlflags=\-shared \-Wl\,\-z\,relro \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.24 \-Darchlib=/usr/lib/x86\_64\-linux\-gnu/perl/5\.24 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/x86\_64\-linux\-gnu/perl5/5\.24 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.24\.1 \-Dsitearch=/usr/local/lib/x86\_64\-linux\-gnu/perl/5\.24\.1 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Dusesitecustomize \-Duse64bitint \-Dman1ext=1 \-Dman3ext=3perl
\-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-Ui\_libutil \-Uversiononly \-DDEBUGGING=\-g \-Doptimize=\-O2 \-dEs \-Duseshrplib \-Dlibperl=libperl\.so\.5\.24\.1'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='x86\_64\-linux\-gnu\-gcc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fwrapv \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fwrapv \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='6\.3\.0 20170221'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678\, doublekind=3
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16\, longdblkind=3
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='x86\_64\-linux\-gnu\-gcc'\, ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/6/include\-fixed /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=libc\-2\.24\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.24
&nbsp;    gnulibc\_version='2\.24'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib \-fstack\-protector\-strong'

Locally applied patches&#8203;:
&nbsp;    DEBPKG&#8203;:debian/cpan\_definstalldirs \- Provide a sensible INSTALLDIRS default for modules installed from CPAN\.
&nbsp;    DEBPKG&#8203;:debian/db\_file\_ver \- https://bugs.debian.org/340047 Remove overly restrictive DB\_File version check\.
&nbsp;    DEBPKG&#8203;:debian/doc\_info \- Replace generic man\(1\) instructions with Debian\-specific information\.
&nbsp;    DEBPKG&#8203;:debian/enc2xs\_inc \- https://bugs.debian.org/290336 Tweak enc2xs to follow symlinks and ignore missing @&#8203;INC directories\.
&nbsp;    DEBPKG&#8203;:debian/errno\_ver \- https://bugs.debian.org/343351 Remove Errno version check due to upgrade problems with long\-running processes\.
&nbsp;    DEBPKG&#8203;:debian/libperl\_embed\_doc \- https://bugs.debian.org/186778 Note that libperl\-dev package is required for embedded linking
&nbsp;    DEBPKG&#8203;:fixes/respect\_umask \- Respect umask during installation
&nbsp;    DEBPKG&#8203;:debian/writable\_site\_dirs \- Set umask approproately for site install directories
&nbsp;    DEBPKG&#8203;:debian/extutils\_set\_libperl\_path \- EU&#8203;:MM&#8203;: set location of libperl\.a under /usr/lib
&nbsp;    DEBPKG&#8203;:debian/no\_packlist\_perllocal \- Don't install \.packlist or perllocal\.pod for perl or vendor
&nbsp;    DEBPKG&#8203;:debian/fakeroot \- Postpone LD\_LIBRARY\_PATH evaluation to the binary targets\.
&nbsp;    DEBPKG&#8203;:debian/instmodsh\_doc \- Debian policy doesn't install \.packlist files for core or vendor\.
&nbsp;    DEBPKG&#8203;:debian/ld\_run\_path \- Remove standard libs from LD\_RUN\_PATH as per Debian policy\.
&nbsp;    DEBPKG&#8203;:debian/libnet\_config\_path \- Set location of libnet\.cfg to /etc/perl/Net as /usr may not be writable\.
&nbsp;    DEBPKG&#8203;:debian/mod\_paths \- Tweak @&#8203;INC ordering for Debian
&nbsp;    DEBPKG&#8203;:debian/prune\_libs \- https://bugs.debian.org/128355 Prune the list of libraries wanted to what we actually need\.
&nbsp;    DEBPKG&#8203;:fixes/net\_smtp\_docs \- \[rt\.cpan\.org \#36038\] https://bugs.debian.org/100195 Document the Net&#8203;::SMTP 'Port' option
&nbsp;    DEBPKG&#8203;:debian/perlivp \- https://bugs.debian.org/510895 Make perlivp skip include directories in /usr/local
&nbsp;    DEBPKG&#8203;:debian/deprecate\-with\-apt \- https://bugs.debian.org/747628 Point users to Debian packages of deprecated core modules
&nbsp;    DEBPKG&#8203;:debian/squelch\-locale\-warnings \- https://bugs.debian.org/508764 Squelch locale warnings in Debian package maintainer scripts
&nbsp;    DEBPKG&#8203;:debian/skip\-upstream\-git\-tests \- Skip tests specific to the upstream Git repository
&nbsp;    DEBPKG&#8203;:debian/patchlevel \- https://bugs.debian.org/567489 List packaged patches for 5\.24\.1\-2 in patchlevel\.h
&nbsp;    DEBPKG&#8203;:debian/skip\-kfreebsd\-crash \- https://bugs.debian.org/628493 \[perl \#96272\] Skip a crashing test case in t/op/threads\.t on GNU/kFreeBSD
&nbsp;    DEBPKG&#8203;:fixes/document\_makemaker\_ccflags \- https://bugs.debian.org/628522 \[rt\.cpan\.org \#68613\] Document that CCFLAGS should include $Config\{ccflags\}
&nbsp;    DEBPKG&#8203;:debian/find\_html2text \- https://bugs.debian.org/640479 Configure CPAN&#8203;::Distribution with correct name of html2text
&nbsp;    DEBPKG&#8203;:debian/perl5db\-x\-terminal\-emulator\.patch \- https://bugs.debian.org/668490 Invoke x\-terminal\-emulator rather than xterm in perl5db\.pl
&nbsp;    DEBPKG&#8203;:debian/cpan\-missing\-site\-dirs \- https://bugs.debian.org/688842 Fix CPAN&#8203;::FirstTime defaults with nonexisting site dirs if a parent is writable
&nbsp;    DEBPKG&#8203;:fixes/memoize\_storable\_nstore \- \[rt\.cpan\.org \#77790\] https://bugs.debian.org/587650 Memoize&#8203;::Storable&#8203;: respect 'nstore' option not respected
&nbsp;    DEBPKG&#8203;:debian/regen\-skip \- Skip a regeneration check in unrelated git repositories
&nbsp;    DEBPKG&#8203;:debian/makemaker\-pasthru \- https://bugs.debian.org/758471 Pass LD settings through to subdirectories
&nbsp;    DEBPKG&#8203;:debian/makemaker\-manext \- https://bugs.debian.org/247370 Make EU&#8203;::MakeMaker honour MANnEXT settings in generated manpage headers
&nbsp;    DEBPKG&#8203;:debian/devel\-ppport\-reproducibility \- https://bugs.debian.org/801523 Sort the list of XS code files when generating RealPPPort\.xs
&nbsp;    DEBPKG&#8203;:debian/encode\-unicode\-bom\-doc \- https://bugs.debian.org/798727 Document Debian backport of Encode&#8203;::Unicode fix
&nbsp;    DEBPKG&#8203;:debian/kfreebsd\-softupdates \- https://bugs.debian.org/796798 Work around Debian Bug\#796798
&nbsp;    DEBPKG&#8203;:fixes/autodie\-scope \- https://bugs.debian.org/798096 Fix a scoping issue with "no autodie" and the "system" sub
&nbsp;    DEBPKG&#8203;:fixes/crosscompile\-no\-targethost \- \[23695c0\] \[perl \#127234\] Fix the Configure escape with usecrosscompile but no targethost
&nbsp;    DEBPKG&#8203;:fixes/memoize\-pod \- \[rt\.cpan\.org \#89441\] Fix POD errors in Memoize
&nbsp;    DEBPKG&#8203;:fixes/ok\-pod \- Added encoding for pod\.
&nbsp;    DEBPKG&#8203;:debian/hurd\-softupdates \- https://bugs.debian.org/822735 Fix t/op/stat\.t failures on hurd
&nbsp;    DEBPKG&#8203;:fixes/nntp\_docs \- https://bugs.debian.org/51962 Net&#8203;::NNTP&#8203;: Correct innd/nnrpd confusion in relation to Reader option
&nbsp;    DEBPKG&#8203;:fixes/math\_complex\_doc\_great\_circle \- https://bugs.debian.org/697567 \[rt\.cpan\.org \#114104\] Math&#8203;::Trig&#8203;: clarify definition of great\_circle\_midpoint
&nbsp;    DEBPKG&#8203;:fixes/math\_complex\_doc\_see\_also \- https://bugs.debian.org/697568 \[rt\.cpan\.org \#114105\] Math&#8203;::Trig&#8203;: add missing SEE ALSO
&nbsp;    DEBPKG&#8203;:fixes/math\_complex\_doc\_angle\_units \- https://bugs.debian.org/731505 \[rt\.cpan\.org \#114106\] Math&#8203;::Trig&#8203;: document angle units
&nbsp;    DEBPKG&#8203;:fixes/cpan\_web\_link \- https://bugs.debian.org/367291 CPAN&#8203;: Add link to main CPAN web site
&nbsp;    DEBPKG&#8203;:fixes/time\_piece\_doc \- https://bugs.debian.org/817925 Time&#8203;::Piece&#8203;: Improve documentation for add\_months and add\_years
&nbsp;    DEBPKG&#8203;:fixes/perlbug\-refactor \- https://bugs.debian.org/822463 \[perl \#128020\] perlbug&#8203;: Refactor duplicated file reading code
&nbsp;    DEBPKG&#8203;:fixes/perlbug\-linewrap \- https://bugs.debian.org/822463 \[perl \#128020\] perlbug&#8203;: wrap overly long lines
&nbsp;    DEBPKG&#8203;:fixes/hurd\_sigaction \- https://bugs.debian.org/825016 \[d54f4ed\] ext/POSIX/t/sigaction\.t&#8203;: Skip uid and pid tests on GNU/Hurd
&nbsp;    DEBPKG&#8203;:fixes/hurd\_hints \- \[4694301\] https://bugs.debian.org/825020 \[perl \#128279\] Modify hints for Hurd per Debian ticket 825020\.
&nbsp;    DEBPKG&#8203;:fixes/extutils\-parsexs\-reproducibility \- \[perl \#128517\] https://bugs.debian.org/829296 Make the output of ExtUtils&#8203;::ParseXS reproducible
&nbsp;    DEBPKG&#8203;:debian/CVE\-2016\-1238/sitecustomize\-in\-etc \- Look for sitecustomize\.pl in /etc/perl rather than sitelib on Debian systems
&nbsp;    DEBPKG&#8203;:debian/CVE\-2016\-1238/test\-suite\-without\-dot \- \[perl \#127810\] Patch unit tests to explicitly insert "\." into @&#8203;INC when needed\.
&nbsp;    DEBPKG&#8203;:debian/CVE\-2016\-1238/eumm\-without\-dot \- \[perl \#127810\] Add PERL\_USE\_UNSAFE\_INC support to EU&#8203;::MM for fortify\_inc support\.
&nbsp;    DEBPKG&#8203;:debian/CVE\-2016\-1238/cpan\-without\-dot \- \[perl \#127810\] Set PERL\_USE\_UNSAFE\_INC for cpan usage
&nbsp;    DEBPKG&#8203;:debian/document\_inc\_removal \- Document in perlvar that we remove '\.' from @&#8203;INC by default
&nbsp;    DEBPKG&#8203;:fixes/extutils\_makemaker\_reproducible \- https&#8203;://bugs\.debian\.org/835815 https://bugs.debian.org/834190 Make perllocal\.pod files reproducible
&nbsp;    DEBPKG&#8203;:debian/CVE\-2016\-1238/remove\-inc\-test \- Remove test for '\.' in @&#8203;INC as it might not be
&nbsp;    DEBPKG&#8203;:debian/customized \- Update customized\.dat for files patched in Debian
&nbsp;    DEBPKG&#8203;:fixes/file\_path\_hurd\_errno \- File\-Path&#8203;: Fix test failure in Hurd due to hard\-coded ENOENT
&nbsp;    DEBPKG&#8203;:debian/hppa\_op\_optimize\_workaround \- https://bugs.debian.org/838613 Temporarily lower the optimization of op\.c on hppa due to gcc\-6 problems
&nbsp;    DEBPKG&#8203;:fixes/test\-builder\-warning \- https://bugs.debian.org/840968 Silence a 'used only once' warning in Test&#8203;::Builder
&nbsp;    DEBPKG&#8203;:fixes/longdblinf\-randomness \- \[dd68853\] \[perl \#130133\] https://bugs.debian.org/844752 Configure&#8203;: fix garbage filtering with 80\-bit long doubles
&nbsp;    DEBPKG&#8203;:debian/installman\-utf8 \- https://bugs.debian.org/840211 Generate man pages with UTF\-8 characters
&nbsp;    DEBPKG&#8203;:fixes/list\_assign\_leak \- \[1050723\] \[perl \#130766\] https://bugs.debian.org/855064 avoid a leak in list assign from/to magic values
&nbsp;    DEBPKG&#8203;:fixes/perlfunc\_inc\_doc \- \[a03e9f8\] https://bugs.debian.org/839536 \[perl \#130832\] Documentation fixes for '\.' possibly no longer being in @&#8203;INC



---
@&#8203;INC for perl 5\.24\.1&#8203;:
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/x86\_64\-linux\-gnu/perl/5\.24\.1
&nbsp;    /usr/local/share/perl/5\.24\.1
&nbsp;    /usr/lib/x86\_64\-linux\-gnu/perl5/5\.24
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/x86\_64\-linux\-gnu/perl/5\.24
&nbsp;    /usr/share/perl/5\.24
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    /usr/lib/x86\_64\-linux\-gnu/perl\-base



---
Environment for perl 5\.24\.1&#8203;:
&nbsp;    HOME=/home/rbalint
&nbsp;    LANG=C
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LC\_ADDRESS=hu\_HU\.UTF\-8
&nbsp;    LC\_ALL=en\_US\.UTF\-8
&nbsp;    LC\_IDENTIFICATION=hu\_HU\.UTF\-8
&nbsp;    LC\_MEASUREMENT=hu\_HU\.UTF\-8
&nbsp;    LC\_MONETARY=hu\_HU\.UTF\-8
&nbsp;    LC\_NAME=hu\_HU\.UTF\-8
&nbsp;    LC\_NUMERIC=hu\_HU\.UTF\-8
&nbsp;    LC\_PAPER=hu\_HU\.UTF\-8
&nbsp;    LC\_TELEPHONE=hu\_HU\.UTF\-8
&nbsp;    LC\_TIME=hu\_HU\.UTF\-8
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/lib/ccache&#8203;:/home/rbalint/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/local/games&#8203;:/usr/games
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From rbalint@ubuntu.com

Apparently my original report was too detailed\, pasting the interesting parts&#8203;:

When building nama Perl crashes in tests\.

Originally reported to Debian BTS&#8203;:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=839218


This looks like a problem in Perl itself possibly causing random crashes
elsewhere\, too\.

Reproduction with the perl\-debug package shows the internal assertion&#8203;:

sudo apt\-get install perl\-debug
sudo apt\-get build\-dep nama
apt source nama
cd nama\-\*
dpkg\-buildpackage
\.\.\.
ok 57 \- set JACK client as input
sax&#8203;: JACK source port is sax\_in\. Make connections manually\.

Track sax&#8203;: source set to JACK manual port sax\_in
Failed 69/126 subtests
t/13\_io\.t \.\.\.\.\.\.\.
1\.\.1
ok 1 \- use Audio&#8203;::Nama&#8203;::IO;
ok
\.\.

PERL\_DL\_NONLAZY=1 PERL\_USE\_UNSAFE\_INC=1 "/usr/bin/debugperl" \\
"\-MExtUtils&#8203;::Command&#8203;::MM" "\-MTest&#8203;::Harness" "\-e" " test\_harness\(1\, \\
'inc'\, 'blib/lib'\, 'blib/arch'\)" t/12\_nama\.t

\.\.\.

ok 57 \- set JACK client as input
sax&#8203;: JACK source port is sax\_in\. Make connections manually\.
Track sax&#8203;: source set to JACK manual port sax\_in
debugperl&#8203;: sv\.c&#8203;:6438&#8203;: Perl\_sv\_clear&#8203;: Assertion \`SvTYPE\(sv\) \!=
\(svtype\)SVTYPEMASK' failed\.
Failed 69/126 subtests
\.\.\.

Address Sanitizer shows the crash with non\-debug perl&#8203;:

LD\_PRELOAD=libasan\.so\.3 PERL\_DL\_NONLAZY=1 PERL\_USE\_UNSAFE\_INC=1 \\
"/usr/bin/perl" "\-MExtUtils&#8203;::Command&#8203;::MM" "\-MTest&#8203;::Harness" "\-e" "
test\_harness\(1\, 'inc'\, 'blib/lib'\, 'blib/arch'\)" t/12\_nama\.t
\.\.\.
sax&#8203;: JACK source port is sax\_in\. Make connections manually\.
Track sax&#8203;: source set to JACK manual port sax\_in
ASAN&#8203;:DEADLYSIGNAL
=================================================================
==13877==ERROR&#8203;: AddressSanitizer&#8203;: SEGV on unknown address 0x00ff00000012
\(pc 0x559191b15fd8 bp 0x62100013f578 sp 0x7fffaa838b00 T0\)
&nbsp;    \#0 0x559191b15fd7  \(/usr/bin/perl\+0xd9fd7\)
&nbsp;    \#1 0x559191b1616d in Perl\_sv\_unmagic \(/usr/bin/perl\+0xda16d\)
&nbsp;    \#2 0x559191b1513a in Perl\_sv\_clear \(/usr/bin/perl\+0xd913a\)
&nbsp;    \#3 0x559191b15a8f in Perl\_sv\_free2 \(/usr/bin/perl\+0xd9a8f\)
&nbsp;    \#4 0x559191b4624e in Perl\_leave\_scope \(/usr/bin/perl\+0x10a24e\)
&nbsp;    \#5 0x559191b49707  \(/usr/bin/perl\+0x10d707\)
&nbsp;    \#6 0x559191b4ebab in Perl\_die\_unwind \(/usr/bin/perl\+0x112bab\)
&nbsp;    \#7 0x559191aea518 in Perl\_vcroak \(/usr/bin/perl\+0xae518\)
&nbsp;    \#8 0x559191aeafb3 in Perl\_croak \(/usr/bin/perl\+0xaefb3\)
&nbsp;    \#9 0x559191b1d592 in Perl\_sv\_setsv\_flags \(/usr/bin/perl\+0xe1592\)
&nbsp;    \#10 0x559191b08ca9 in Perl\_pp\_sassign \(/usr/bin/perl\+0xccca9\)
&nbsp;    \#11 0x559191b08605 in Perl\_runops\_standard \(/usr/bin/perl\+0xcc605\)
&nbsp;    \#12 0x559191a8e6c8 in perl\_run \(/usr/bin/perl\+0x526c8\)
&nbsp;    \#13 0x559191a6787c in main \(/usr/bin/perl\+0x2b87c\)
&nbsp;    \#14 0x7fa31be502b0 in \_\_libc\_start\_main
\(/lib/x86\_64\-linux\-gnu/libc\.so\.6\+0x202b0\)
&nbsp;    \#15 0x559191a678b9 in \_start \(/usr/bin/perl\+0x2b8b9\)

AddressSanitizer can not provide additional info\.
SUMMARY&#8203;: AddressSanitizer&#8203;: SEGV \(/usr/bin/perl\+0xd9fd7\)
==13877==ABORTING
Dubious\, test returned 1 \(wstat 256\, 0x100\)
Failed 69/126 subtests

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From rbalint@ubuntu.com


Niko Tyni found a workaround\, but Perl itself would still be buggy&#8203;:

> While I haven't made it to the bottom of this yet\, I note that there's
> this code in lib/Audio/Nama/Effect\.pm&#8203;:
>
>          my $perl\_version = $^V;
>          my \($minor\_version\) = $perl\_version =~ /^v5\\\.\(\\d\+\)/;
>          my @&#8203;new\_list = grep  \{ $\_ ne $id  \} @&#8203;ops\_list;
>          \#say "new\_list&#8203;: @&#8203;new\_list";
>          if \($minor\_version \<= 14\)
>               \{    $track\->\{ops\}   = \[ @&#8203;new\_list  \] \}
>          else \{ @&#8203;\{ $track\->\{ops\} \} =    @&#8203;new\_list    \}
>
> and if I patch that to take the \<= 14 code path\, the segfault goes away
> here and all tests pass\. What's the history behind this version check?

Funny story\, it was added to work around a segfault in older perl&#8203;:
https://github.com/bolangi/nama/commit/8ecdbd5ad0b8ac67b7e4adab2ee1100fbf29aa1f

The line looks very much like a corner case for the assignment
reimplementation&#8203;:


 https://github.com/Perl/perl5/commit/a5f48505593c7e1ca478de383e24d5cc2541f3ca

&nbsp;       re\-implement OPpASSIGN\_COMMON mechanism

 This commit almost completely replaces the current mechanism
 for detecting and handing common vars in list assignment\, e\.g\.
 \.\.\.


The issue was also reported against nama\, but I could not find a way
to reassign it to perl&#8203;:
https://rt.cpan.org/Public/Bug/Display.html?id=109852

**Comment by p5pRT** [comment]
### From @ntyni

On Fri\, Mar 24\, 2017 at 07&#8203;:41&#8203;:40AM \-0700\, Balint Reczey wrote&#8203;:

> ok 57 \- set JACK client as input
> sax&#8203;: JACK source port is sax\_in\. Make connections manually\.
> Track sax&#8203;: source set to JACK manual port sax\_in
> debugperl&#8203;: sv\.c&#8203;:6438&#8203;: Perl\_sv\_clear&#8203;: Assertion \`SvTYPE\(sv\) \!=
> \(svtype\)SVTYPEMASK' failed\.
> Failed 69/126 subtests

I reduced this down to

&nbsp;  my $a = \[ 0\, 1 \];
&nbsp;  
&nbsp;  sub f \{
&nbsp;      my $arg = shift;
&nbsp;      my @&#8203;a1 = @&#8203;$a;
&nbsp;      @&#8203;$a = @&#8203;a1;
&nbsp;      return\(\);
&nbsp;  \}
&nbsp;  
&nbsp;  map\{ f\($\_\) \} @&#8203;$a;

which gives

&nbsp;  panic&#8203;: attempt to copy freed scalar 55aa6f0921c0 to 55aa6f092bb0 at \- line 4\.
&nbsp;  debugperl&#8203;: sv\.c&#8203;:6438&#8203;: Perl\_sv\_clear&#8203;: Assertion \`SvTYPE\(sv\) \!= \(svtype\)SVTYPEMASK' failed\.
&nbsp;  zsh&#8203;: abort \(core dumped\)  debugperl
 
So another stack\-not\-refcounted issue?

\-\- 
Niko Tyni   ntyni@&#8203;debian\.org

**Comment by p5pRT** [comment]
### From zefram@fysh.org

Niko Tyni wrote&#8203;:
>So another stack\-not\-refcounted issue?

Yes\.

\-zefram
